layerfetcher/layers.go

```diff
@@ -322,67 +322,70 @@ func (l *Logic) receiveLayerHash(ctx context.Context, id types.LayerID, p p2ppee
 		return
 	}
 
-	l.log.Info("got hashes %v for layer, now querying peers", len(hashes))
+	l.log.Info("got %v hashes for layer, now querying peers", len(hashes))
 
 	// send a request to get blocks from a single peer if multiple peers declare same hash per layer
 	// if the peers fails to respond request will be sen to next peer in line
 	//todo: think if we should aggregate or ask from multiple peers to have some redundancy in requests
-	for hash, peer := range hashes {
+	for hash, peers := range hashes {
 		if hash == emptyHash {
 			l.receiveBlockHashes(ctx, id, nil, len(hashes), ErrZeroLayer)
 			continue
 		}
-		//build receiver function
+		// build receiver function
 		receiveForPeerFunc := func(data []byte) {
 			l.receiveBlockHashes(ctx, id, data, len(hashes), nil)
 		}
-		remainingPeers := 0
 		errFunc := func(err error) {
 			l.receiveBlockHashes(ctx, id, nil, len(hashes), err)
 		}
-		err := l.net.SendRequest(ctx, LayerBlocksMsg, hash.Bytes(), peer[remainingPeers], receiveForPeerFunc, errFunc)
+		err := l.net.SendRequest(ctx, LayerBlocksMsg, hash.Bytes(), peers[0], receiveForPeerFunc, errFunc)
 		if err != nil {
 			l.receiveBlockHashes(ctx, id, nil, len(hashes), err)
 		}
 	}
 
 }
 
+func (l *Logic) determinePromiseResult(layer types.LayerID, errs []error) *LayerPromiseResult {
+	hasZeroBlockHash := false
+	for _, e := range errs {
+		if e == nil {
+			// at least one layer hash contains blocks. not a zero block layer
+			return &LayerPromiseResult{Layer: layer, Err: nil}
+		}
+		if e == ErrZeroLayer {
+			hasZeroBlockHash = true
+		}
+	}
+	if hasZeroBlockHash {
+		// all other non-empty layer hashes returned errors. use the best information we've got
```

Comment:
Why "all other"? Either this comment is misleading or else I'm misunderstanding something. Wouldn't _one_ such result be enough to trigger this logic?

Reply:
i mean all other hashes that's not empty hash returned errors.

because we want to return a valid result if there is one. hence the early return at line 355.
so seeing one such result (ErrZeroLayer) isn't enough. we need to make sure there is no valid result before we determine this is an zero-block layer. see test case Test_notifyLayerPromiseResult_OneHasBlockData